### DISCUSSION TITLE
Super admin and another admins

### DISCUSSION THREAD

**Discussion by neoacevedo** [spatie/laravel-permission#1868] [Q&A] [answered]
I have defined a super admin role and I have another roles. All them lands on /dashboard route.

According with https://spatie.be/docs/laravel-permission/v5/basic-usage/super-admin, I have to setup the gate into the AuthServiceProvider boot method for super admin. It's ok, but, other roles will run in a forbidden status when they land in dashboard.

How can I to allow the dashboard for all roles?

I forgot: in the __construct method for AdminController, where I have defined the dashboard route, I have setup the middleware but it seems not work.

**Comment by erikn69** [comment]
> Y la regla donde dice que no se puede preguntar en español, dónde está?

Entonces porque preguntas en ingles? 😄

**Reply by neoacevedo** [reply]
Acto reflejo.

**Comment by erikn69** [comment, marked as answer]
>How can I to allow the dashboard for all roles?

It looks easy to me, only use request
```php
Gate::before(function ($user, $ability) {
    return (request()->is('dashboard') && $user->roles->count()) || $user->hasRole('Super Admin') ? true : null;
});
```

**Reply by erikn69** [reply]
Did you try this

>Alternatively, you can separate multiple roles or permission with a | (pipe) character:

```php
$this->middleware([
     'auth:admin',
     'permission:admins|audit|log|roles|permissions|payment|all'
]);
```
I don't use middleware, I'm just guessing according [docs](https://spatie.be/docs/laravel-permission/v5/basic-usage/middleware#package-middleware) that I suppose you read

**Reply by neoacevedo** [reply]
It's similar, I get extactly the same result.

I checked this issue: https://github.com/spatie/laravel-permission/issues/1354#issuecomment-583658126 and I adapted to my current issue what worked for me; but, why your bellow code isn't working as I wanted to?

**Reply by erikn69** [reply]
Weird, it should be successful after finding the first permission allowed and never get error on `all` permission because it is at the end https://github.com/spatie/laravel-permission/blob/2512d5a22aaae0bab8d62a02e502a603776da9da/src/Middlewares/PermissionMiddleware.php#L22-L26

**Reply by neoacevedo** [reply]
Yes, and, in another controller, I updated the role admin for list (index) and edit the users. The permisisons are:
_users/index, users/create, users/edit, users/delete_ but, no matter the permissions, this user is able to edit or delete, and in the gate I see:

```
success

array:4 [▼
  "ability" => "users/index"
  "result" => true
  "admin" => 18
  "arguments" => "[]"
]
```

Why? I don't have a clue.

**Reply by erikn69** [reply]
If you are doing this (`permission:admins|audit|log|roles|permissions|payment|all`) with the controller if user have one permission allowed, he can access all the routes, example if have index permission can edit, delete, etc

**Reply by neoacevedo** [reply]
Ok, then... it wouldn't have any sense to have roles with permissions and to use any middleware with permission:admins|audit|log|roles|permissions|payment|all` because with any one of the permission listed, the user will be able to do anything.

I think the permissions validation should be grouped and not allowed on first permission allowed, more or less, via controller@method or via URL. I mean, a user with rol with permissions only to see log wouldn't be allowed to access to any other permission except log.

**Reply by erikn69** [reply]
I using `callAction` for that on a trait for controller, something like this:
```php
public function callAction($method, $parameters) {
     if( $ability = $this->getAbility($method) ) {
         $this->authorize($ability);
     }
     return parent::callAction($method, $parameters);
}
```
**Source :** [Authorizable.php](https://github.com/saqueib/roles-permissions-laravel/blob/master/app/Authorizable.php)

**Reply by neoacevedo** [reply]
I will check it.

**Reply by neoacevedo** [reply]
Within the method `getAbility`, if I change to:

```php
        $routeName = explode('/', request()->route()->uri);
        $allowed = auth()->user()->can($method);
        return $allowed? $routeName[0] . "/" . $method : null;
```

It will allow the action, even if `$allowed` is false, but if I let the method as is, it works. Do you have any explanation about it?

**Reply by erikn69** [reply]
>Do you have any explanation about it?

I have basic knowledge, no idea 😄
It's not my code, and i'm not an collaborator of the package, i'm only trying to help

**Reply by neoacevedo** [reply]
jaja Ok don't worry. Thank you anyway. I will try to solve by myself now.

**Reply by erikn69** [reply]
Here are the docs of that code: [easy-roles-and-permissions](http://qcode.in/easy-roles-and-permissions-in-laravel-5-4/)

**Comment by neoacevedo** [comment]
Good, works for dashboard. 

Now, I see not works for the other routes I need. I think it is related with the permissions themselves.

I get, for another routes, the same behavior as described in https://github.com/spatie/laravel-permission/issues/1027 

I will try to be as clearest as I can. I have some roles: super admin, admin, accountant.

I don't want to set the permissions on each view, I think it breaks the MVC pattron, and I don't want at the moment to setup the permissions in each method, and setting the permissions in the routes file isn't a good idea because there is a lot of routes, then, I use the controller constructor method for setup the middleware:

```php
$this->middleware([
            'auth:admin',
            'permission:admins',
            'permission:audit',
            "permission:log",
            'permission:roles',
            'permission:permissions',
            'permission:payment',
            "permission:all"
        ]);
```

The super admin can do everything (all). The admin rol doesn't have some permissions, for instance, this rol have log, roles and permissions as asigned permissions but it has the admins, audit and payment permissions; the accountant only has the payment permission. 

My problem rises when, the user with admin role tries to access to any allowed route (according with the permissions) in the controller:

```
array:4 [▼
  "ability" => "admins"
  "result" => true
  "admin" => 16
  "arguments" => "[]"
]

success

array:4 [▼
  "ability" => "audit"
  "result" => true
  "admin" => 16
  "arguments" => "[]"
]

success

array:4 [▼
  "ability" => "log"
  "result" => true
  "admin" => 16
  "arguments" => "[]"
]

error

array:4 [▼
  "ability" => "all"
  "result" => null
  "admin" => 16
  "arguments" => "[]"
]

```

The middleware checks this role and it sees the "all" permissions is not asigned and raises the error 403. 

I'd like to know for suggestions to achieve what I need.

Thank you.
